packages/flutter/lib/src/widgets/overlay.dart

```diff
@@ -476,223 +489,319 @@ class OverlayState extends State<Overlay> with TickerProviderStateMixin {
   }
 }
 
-/// A widget that has one [onstage] child which is visible, and one or more
-/// [offstage] widgets which are kept alive, and are built, but are not laid out
-/// or painted.
+/// Special version of a [Stack], that doesn't layout and render the first
+/// [skipCount] children.
 ///
-/// The onstage widget must be a [Stack].
-///
-/// For convenience, it is legal to use [Positioned] widgets around the offstage
-/// widgets.
-class _Theatre extends RenderObjectWidget {
+/// The first [skipCount] children are considered "offstage".
+class _Theatre extends MultiChildRenderObjectWidget {
   _Theatre({
-    this.onstage,
-    @required this.offstage,
-  }) : assert(offstage != null),
-       assert(!offstage.any((Widget child) => child == null));
-
-  final Stack onstage;
+    Key key,
+    this.skipCount = 0,
+    List<Widget> children = const <Widget>[],
+  }) : assert(skipCount != null),
+       assert(skipCount >= 0),
+       assert(children != null),
+       assert(children.length >= skipCount),
+       super(key: key, children: children);
 
-  final List<Widget> offstage;
+  final int skipCount;
 
   @override
   _TheatreElement createElement() => _TheatreElement(this);
 
   @override
-  _RenderTheatre createRenderObject(BuildContext context) => _RenderTheatre();
+  _RenderTheatre createRenderObject(BuildContext context) {
+    return _RenderTheatre(
+      skipCount: skipCount,
+      textDirection: Directionality.of(context),
+    );
+  }
+
+  @override
+  void updateRenderObject(BuildContext context, _RenderTheatre renderObject) {
+    renderObject
+      ..skipCount = skipCount
+      ..textDirection = Directionality.of(context);
+  }
+
+  @override
+  void debugFillProperties(DiagnosticPropertiesBuilder properties) {
+    super.debugFillProperties(properties);
+    properties.add(IntProperty('skipCount', skipCount));
+  }
 }
 
-class _TheatreElement extends RenderObjectElement {
-  _TheatreElement(_Theatre widget)
-    : assert(!debugChildrenHaveDuplicateKeys(widget, widget.offstage)),
-      super(widget);
+class _TheatreElement extends MultiChildRenderObjectElement {
+  _TheatreElement(_Theatre widget) : super(widget);
 
   @override
   _Theatre get widget => super.widget as _Theatre;
 
   @override
   _RenderTheatre get renderObject => super.renderObject as _RenderTheatre;
 
-  Element _onstage;
-  static final Object _onstageSlot = Object();
+  @override
+  void debugVisitOnstageChildren(ElementVisitor visitor) {
+    assert(children.length >= widget.skipCount);
+    children.skip(widget.skipCount).forEach(visitor);
```

Comment:
Similar concern - I think we at least want an assert here that all the widgets we're skipping are in fact offstage.

Reply:
It's the other way around, though: The fact that we are skipping them makes them offstage. There's no other marker that marks them as offstage that we could assert on.